### PR TITLE
Refactor workflow to support PERs and ARs

### PR DESCRIPTION
**Pull request by Crell**
Let's finally define a process for "living specs", which I have here called PHP Evolving Recommendations (PERs).  This now involves a number of moving parts, which largely come down to refactoring bits of the PSR process out to separate pieces and reusing them more formally.

# Changelog

* Simplify voting somewhat by defining two base vote types: Approval (Yes/No) and Decision (multiple choice).
* Establish a process for Implicit Approval by the Core Committee to minimize pointless voting.
* Move the Working Group process out of the PSR workflow into its own section under Mission and Structure, complete with full lifecycle personnel management.
* Establish two "levels" of Working Group.
* Refactor the PSR workflow to take advantage of the above changes.
* Define a new workflow for PERs using the above independent bits.
* Explicitly list the types of artifact we produce.
* Defines the role of "Maintainer" for artifacts, including utility libraries plus PSR/PER output.
* Reorders the bylaws to put PERs up front by PSRs.

As a practical matter, basically nothing changes for PSRs from today.  The changes look larger because it's lots of refactoring out to base classes and composed objects rather than one big god object, but the unit tests for the PSR process should still all pass.  If, you know, that was a thing... :smile:

### CODE REVIEW

File: bylaws/009-per-workflow.md
Comment (Crell) [inline bylaws/009-per-workflow.md:83]: I don't like the name for this section.  Please suggest something better.

Comment (benedmunds) [reply bylaws/009-per-workflow.md:75]:
```suggestion
## Expedited PER Working Group
```

**Comment by benedmunds** [comment]
I'm not sure where to put it but it's probably worth a TLDR of the differences between PSRs and PERs without having to dive into the workflow to figure it out.

**Comment by Crell** [comment]
It might make sense to merge the workflows at some point, at least descriptively.  But that's more work so I didn't want to get into it unless we were certain we wanted to go that route. :smile: 

But the tldr is basically what's in the description now.  A PER is the same as a PSR except it gets a name instead of a number, and can have SemVer releases after it's approved.  3/4 of the new PER page is a copy-paste from the PSR page.

File: bylaws/009-per-workflow.md
Comment (benedmunds) [inline bylaws/009-per-workflow.md:69]: Would an "Accepted" PER be marked as "Abandoned" if it's WG dissolves?  Or would it forever remain "Accepted"?

**Comment by benedmunds** [comment]
> But the tldr is basically what's in the description now
Ah yes, I didn't mean for our purposes here, I was thinking of a TLDR for readers.

File: bylaws/009-per-workflow.md
Comment (Crell) [reply bylaws/009-per-workflow.md:69]: The PER would remain accepted, I presume. I suppose we probably do want to have a process to officially deprecate and abandon a PER's artifacts.

File: bylaws/009-per-workflow.md
Comment (alekitto) [inline bylaws/009-per-workflow.md:21]: Naming convention: is there a reason for the descriptive names to be in camel case?
I thought that these names were meant for humans, allowing spaces to be used (ex `Coding Standards`). Defining the package name, kebab-case convention should be used.

File: bylaws/003-votes.md
Comment (vdelau) [inline bylaws/003-votes.md:48]: Who is eligible to vote? Quorum is 50% of what group? I assume CC-only for both but I think it should be specified like in the other paragraphs in this bylaw.

**Comment by vdelau** [comment]
Personally I believe there could be a clearer separation between the product (PSR, PER) and the process (WG). I can see a WG maintaining one or more products, and proposing new products where they identify a need or opportunity. Forming a new WG would probably require a vote to establish a WG charter and initial lead ('editor'). The charter should specify the purpose of the WG and the products it inherits (existing PSR's, PER's) and what products it should deliver (newly proposed PSR's and PER's). Introducing a new product also requires a vote and establish which WG is responsible for the product and who the editor/maintainer and sponsor are.

File: bylaws/009-per-workflow.md
Comment (Jean85) [inline bylaws/009-per-workflow.md:51]: I would reword "considered" into something more tangible, along the lines of "this is the moment where the editor will effectively tag the 1.0".

File: bylaws/007-psr-naming-conventions.md
Comment (shadowhand) [inline bylaws/007-psr-naming-conventions.md:9]: Why the discrepancy here? If a PSR namespace is `Psr`, why wouldn't a PER namespace be `Per`?

File: bylaws/007-psr-naming-conventions.md
Comment (shadowhand) [inline bylaws/007-psr-naming-conventions.md:12]: Same question here.

File: bylaws/007-psr-naming-conventions.md
Comment (shadowhand) [inline bylaws/007-psr-naming-conventions.md:20]:
```suggestion
9. Special lightweight utility packages MAY be produced alongside PSRs and PERs, and
```

File: bylaws/009-per-workflow.md
Comment (shadowhand) [inline bylaws/009-per-workflow.md:3]: I like the wording here, and I think it might be helpful to expand on what "ecosystem develops" means; new features added to the language, deprecations, etc. Maybe something as simple as:

> as the PHP ecosystem develops due to new PHP features and deprecations.

File: bylaws/009-per-workflow.md
Comment (shadowhand) [inline bylaws/009-per-workflow.md:61]: Along with my previous comment, I am not sure that "any time" is a good idea here. Perhaps this should be more defined, per SemVer:

1. Bug fixes MAY be tagged by the Editor at any time.
2. Minor versions MUST be declared to the Core Committee and have a 7 day holding period, during which any Core Committee member MAY call for an Acceptance Vote.
3. Major versions MUST be declared to the Core Committee and MUST be put to an Acceptance Vote.

File: bylaws/009-per-workflow.md
Comment (shadowhand) [inline bylaws/009-per-workflow.md:65]:
```suggestion
The Editor of an approved PER is empowered to add or remove members from the Working Group at any time. Any member changes to the Working Group MUST be declared to the Core Committee by a post on the mailing list with a link to a Pull Request that documents the change.
```

File: bylaws/009-per-workflow.md
Comment (shadowhand) [inline bylaws/009-per-workflow.md:83]: This section is confusing. I understand that the intention is for a PSR working group to be converted to a PER working group after the approval of the PSR, but the wording is rather circuitous.

File: bylaws/009-per-workflow.md
Comment (Crell) [reply bylaws/009-per-workflow.md:21]: Not in particular, beyond knee-jerk preference for strings that can be identifiers.  In this case it probably is better to let it be longer.

File: bylaws/007-psr-naming-conventions.md
Comment (Crell) [reply bylaws/007-psr-naming-conventions.md:9]: Mainly because I was using PERs as a place for util libraries to live, and that already has an established `Fig` namespace.

If we split those up into 3 separate tracks, then I don't know off hand if it makes sense for PERs to have their own namespace or not.  Many (eg, coding standards) likely won't even have code artifacts.

File: bylaws/007-psr-naming-conventions.md
Comment (Crell) [reply bylaws/007-psr-naming-conventions.md:20]: Yep, omission.  May change depending on how deep we decide to refactor this. (Per discussion on list.)

File: bylaws/009-per-workflow.md
Comment (Crell) [reply bylaws/009-per-workflow.md:3]: Or just "as the PHP language and ecosystem develop"?

File: bylaws/009-per-workflow.md
Comment (Crell) [reply bylaws/009-per-workflow.md:57]: Mm, possibly.  My concern is that a whole crapload of stuff could be in a given release, and then if the CC objects you have to sort out how to pull out or change selected pieces.

"Ask forgiveness, not permission" makes saying "no" much harder, even in cases where saying no is the right thing to do.  And sometimes, it will be the right thing to do.

File: bylaws/009-per-workflow.md
Comment (Crell) [reply bylaws/009-per-workflow.md:83]: Yeah, there's a lot of conditionals in this process.  The intent is that, for instance, should PSR-5 ever actually get finished, it would be a PSR that defines parsing rules.  Then after it passes it would mutate into a PER WG to define the tag library.  Or, PSR-18 may have produced a util library, but there's little reason to have a separate PER WG to maintain that when there's already the one left over from PSR-7/15/17 and the PSR-18 people want to just dump the PSR-18 code artifacts on the existing WG to handle (which is probably the best course of action).

If we do a deeper refactor as the list seems to be suggesting I will try to flesh this out further.

File: bylaws/009-per-workflow.md
Comment (shadowhand) [reply bylaws/009-per-workflow.md:3]: I like that.

File: bylaws/009-per-workflow.md
Comment (shadowhand) [reply bylaws/009-per-workflow.md:57]: Well I would hope that the people most likely to have strong opinions about a project would be part of the Working Group. I am pretty sure that "notify the CC of any change" is going to be overly burdensome.

**Comment by Crell** [comment]
OK, big big update pushed.  I like this version a lot better.  See the updated description for the full list of meaningful changes.

**Review by PowerKiKi — CHANGES_REQUESTED** [review]
This introduce 47 cases of two spaces after a period (`.  ` instead of `. `), that should be corrected. Because it breaks consistency with existing content. And because it is trying to style Markdown that will be rendered as HTML in most cases anyway, so it is not really worth the effort.

**Comment by shadowhand** [comment]
I agree about the double-space thing, it actually makes my brain skip as I read the raw markdown.

File: bylaws/001-mission-and-structure.md
Comment (shadowhand) [inline bylaws/001-mission-and-structure.md:96]:
```suggestion
### Working Group Management
```

**Comment by shadowhand** [comment]
These changes satisfy my previous concerns.

File: bylaws/001-mission-and-structure.md
Comment (Jean85) [inline bylaws/001-mission-and-structure.md:128]: Please use the dashed form in any place in our documents:
```suggestion
The PHP-FIG produces three primary forms of output artifact that are subject to established workflow processes.  This list is not exhaustive of the activity or publication that the PHP-FIG may engage in.
```

File: bylaws/001-mission-and-structure.md
Comment (Jean85) [inline bylaws/001-mission-and-structure.md:130]: Should we reference the evolution bylaw as the only exception to this "frozen" situation?

**Review by Jean85 — CHANGES_REQUESTED** [review]

File: bylaws/001-mission-and-structure.md
Comment (Crell) [reply bylaws/001-mission-and-structure.md:130]: Mm, probably.  I'll add.

We should probably also add a bunch of cross-links to the bylaws at this point, but that's probably a task for later.

File: bylaws/001-mission-and-structure.md
Comment (Crell) [reply bylaws/001-mission-and-structure.md:128]: Curiously, the bylaws seem to use PHP FIG almost exclusively, not PHP-FIG.  (It's 15 to 0 in the mission-and-structure bylaw right now.)  That may be worth changing, but seems separate from the other changes here.

Comment (Jean85) [reply bylaws/001-mission-and-structure.md:128]: If that's the case, it's a task for us @php-fig/secretaries, with no urgency.

**Review by Jean85 — APPROVED** [review]

**Review by ashnazg — APPROVED** [review]

File: bylaws/001-mission-and-structure.md
Comment (samdark) [inline bylaws/001-mission-and-structure.md:76]: A bit of duplication here. Not that bad but repetitive.

> and has final authority on the group's output

> is the final authority on the output of a Working Group's efforts

File: bylaws/009-naming-conventions.md
Comment (samdark) [inline bylaws/009-naming-conventions.md:9]: Why not `Per`/`per`?

**Review by samdark — APPROVED** [review]

File: bylaws/009-naming-conventions.md
Comment (Crell) [reply bylaws/009-naming-conventions.md:9]: Mainly because `Fig` is already established and in use, and I see little reason to add another namespace for us to manage. :smile: If we were starting over today, I think we'd probably use `fig/` as the namespace for everything, including PSRs.

File: bylaws/009-naming-conventions.md
Comment (zoispag) [inline bylaws/009-naming-conventions.md:10]:
```suggestion
7. There MUST be a package/second-level namespace in relation to the PSR or PER that covers the code.
```

**Review by Jean85 — APPROVED** [review]
Vote is closed, the bylaw change is approved!

Links to the votes:
 * https://groups.google.com/g/php-fig/c/llG3bZ3Z5jo
 * https://groups.google.com/g/php-fig/c/sNHGSClcwB0

**Comment by Jean85** [comment]
@Crell feel free to port the remainder suggested changes in follow-up PRs, since they all seems to be all trivial changes.

**Pull request merged**
